README
Class-Lite version v0.0.1

Nature's great masterpiece, an elephant, 
The only harmless great thing.
-- John Donne

The hashref-based base class that does no more than it must. 

INSTALLATION

To install this module with Module::Build, run the following commands:

    perl Build.PL
    Build
    Build test
    Build install

DEPENDENCIES

perl 5.8.8 or better. 
For Perl module dependencies, please see Build.PL.
This module has no non-core dependencies; however you may need to upgrade 
some core modules if your perl is old. 

LICENSE

Copyright (C) 2013 Xiong Changnian <xiong@cpan.org>

This library and its contents are released under Artistic License 2.0:

(http://www.opensource.org/licenses/artistic-license-2.0.php)


top/Changes
Revision history for Class-Lite

0.1.0   2013-04-12 20:40
    - Initial release.

0.0.0   2013-04-09 18:52
    - Launch project.



lib/Class/Lite.pm
package Class::Lite;
# Choose minimum perl interpreter version; delete the rest.
# Do you want to enforce the bugfix level?
#~ use 5.008008;   # 5.8.8     # 2006  # oldest sane version
#~ use 5.008009;   # 5.8.9     # 2008  # latest 5.8
#~ use 5.010001;   # 5.10.1    # 2009  # say, state, switch
#~ use 5.012003;   # 5.12.5    # 2011  # yada
#~ use 5.014002;   # 5.14.3    # 2012  # pop $arrayref, copy s///r
#~ use 5.016002;   # 5.16.2    # 2012  # __SUB__
use strict;
use warnings;
use version; our $VERSION = qv('v0.1.0');

# Alternate uses
#~ use Devel::Comments '###', ({ -file => 'debug.log' });                   #~

## use
#============================================================================#

#=========# CLASS METHOD
#~ my $self    = My::Class->new(@_);
#
#   Classic hashref-based-object constructor.
#   Passes any arguments to init().
#   
sub new {
    my $class   = shift;
    my $self    = {};
    bless ( $self => $class );
    $self->init(@_);
    return $self;
}; ## new

#=========# OBJECT METHOD
#~ $self->init(@_);
#
#   Abstract method does nothing. Override in your class.
#   
sub init {
    return shift;
}; ## init

#=========# CLASS METHOD
#~ use Class::Lite qw| attr1 attr2 attr3 |;
#~ use Class::Lite qw|             # Simple base class with get/put accessors
#~     attr1
#~     attr2
#~     attr3
#~ |;
#
#   @
#   
sub import {
    no warnings 'uninitialized';
    my $class       = shift;
#~     return unless shift eq q{Class::Lite};
    my $caller      = caller;
    my $bridge      = qq{Class::Lite::$caller};
    ### $class
    ### $bridge
    ### $caller
    
    # In case caller is eager.
    my @args        = $class->fore_import(@_);
    ### @args
    
    # Do most work in the bridge class.    
    eval join qq{\n},
        qq* package $bridge;                                            *,
        qq* our  \@ISA;                                                 *,
        qq* push \@ISA, '$class';                                       *,
        map {
            defined and ! ref and /^[^\W\d]\w*\z/s
                or die "Invalid accessor name '$_'";
              qq* sub get_$_ { return \$_[0]->{$_} };                   *
            . qq* sub put_$_ { \$_[0]->{$_} = \$_[1]; return \$_[0] };  *
        } @args,
    ;
    die "Failed to generate $bridge: $@" if $@;
    
    # Make caller inherit from bridge.
    eval join qq{\n},
        qq* package $caller;                                            *,
        qq* our  \@ISA;                                                 *,
        qq* push \@ISA, '$bridge';                                      *,
    ;
    die "Failed to generate $caller: $@" if $@;
    
    # In case caller must get the last word.
    $class->rear_import(@_);
    
    return 1;    
}; ## import

# Dummy methods do nothing.
sub fore_import { shift; return @_ };
sub rear_import { shift; return @_ };

## END MODULE
1;
#============================================================================#
__END__

=head1 NAME

Class::Lite - Simple base class with get/put accessors

=head1 VERSION

This document describes Class::Lite version v0.1.0

=head1 SYNOPSIS

    package Toy::Class;
    use Class::Lite qw| foo bar baz |;              # make get/put accessors
    
    package Any::Class;
    use Toy::Class;
    my $toy     = Toy::Class->new;
    $toy->init(@_);                                 # does nothing; override
    $toy->put_foo(42);
    my $answer  = $toy->get_foo;
    
    use Class::Lite;                                # no accessors

=head1 DESCRIPTION

=over

I<< Nature's great masterpiece, an elephant, 
The only harmless great thing.  >> 
-- John Donne

=back

The hashref-based base class that does no more than it must. Your 
constructor and accessors are defined in a bridge package so you can 
override them easily. 

=head1 Why?

Computer programmers are clever people who delight in evading restrictions.
Create an L<< inside-out|Class::Std >> (flyweight) class to enforce 
encapsulation and another fellow will L<< hack in|PadWalker >>. The only 
way to win the ancient game of locksmith and lockpick is never to begin. 
If someone misuses your class then it's not your responsibility. 
Hashref-based objects are traditional, well-understood, even expected in 
the Perl world; tools exist with which to work with them. 

Similarly, C<< Class::Lite >> provides no read-only accessors. If your client 
developer wants to alter an attribute he will; you may as well provide a 
method for the purpose. You might warn against the practice by overriding 
the default method: 

    sub put_foo {
        warn q{Please don't write to the 'foo' attribute.};
        my $self    = shift;
        return $self->SUPER::put_foo(@_);
    };

B<< set >> is too similar to B<< get >> in one way, not enough in another. 
Also B<< set >> is one of those heavily overloaded words, like "love" or 
"data", that I prefer to avoid using at all. I say B<< put >> is equally 
short, clearer in intent, not easily misread for B<< get >>; and the first 
character's descender points in the opposite direction.

I eschew single-method C<< foo() >> accessors. 

I have long defined C<< init() >> as a shortcut method to fill up a new 
object; but this is a blatant violation of encapsulation, no matter who 
does it. No more. 

If accessors are defined in your calling package then you will raise a 
warning if you attempt to redefine them; if they are defined in 
C<< Class::Lite >> itself then they will be available to all that inherit 
from it. So your accessors are defined in an intermediate "bridge" package 
generated at compile-time. 

=head1 USE-LINE

    package Toy::Class;
    use Class::Lite qw| foo bar baz |;              # make get/put accessors
    use Class::Lite;                                # no accessors

Makes C<< Class::Lite >> a base class for Toy::Class. If arguments are 
given then simple get and put accessors will be created in caller's 
namespace for each argument. The accessors do no validation. 

B<< This is probably all you need to know. >> Read on if you intend to do tricky stuff in a superclass. 

=head1 INHERITED METHODS 

=head2 import()

    Class::Lite->import(@_);
    A::Subclass->import(@_);

Called by use() as usual and does all the work. Inherited by caller so 
your further subclasses can also take advantage of C<< Class::Lite >> 
features. 

Since this is merely inherited you may define your own C<< import() >> with 
impunity. If you want to have your cake and eat it, too, beware: 

    package Big;
    sub import {
        my $class       = shift;
        # Do specific stuff...
        $class->SUPER::import(@_);
        return 1;
    };
    
    package Tot;
    use Big (@args);

This will not work as you expect! C<< SUPER::import() >> will think Big is 
its C<< caller() >>, which is true. So instead of making Big a parent of 
Tot and defining accessors for Tot; C<< SUPER::import() >> will attempt to 
make Big a parent of itself... at which point the fatal error relieves us 
of further worry. 

=head2 fore_import()

    package Big;
    sub fore_import {
        my $class       = shift;
        my $args        = shift;
        my $hoge        =    $args->{hoge}      // 'default'     ;
        my @accessors   = @{ $args->{accessors} // []           };
        _do_hoge{$hoge};
        return @accessors;
    };
    
    package Tot;
    use Big {
        hoge        => 'piyo',
        accessors   => [qw| chim chum choo |],
    };

To solve the difficulty previously mentioned: Leave C<< import() >> 
untouched and do whatever you like to the use-line argument list in a 
redefined C<< fore_import() >>. Just be sure to return a flat list of 
arguments so C<< import() >> can do its work. 

The default method does nothing and merely returns its arguments. 

=head2 rear_import()

If you just have to get the last word, redefine C<< rear_import() >> 
instead, or also. You'll be passed all the use-line arguments, not just 
what C<< fore_import() >> returned; and your return value will be 
discarded. 

The default method does nothing and merely returns its arguments. 

NOTE that neither of these methods must be employed if all you want to do in your class is override C<< Class::Lite::import() >> completely. 

=head2 new()

    my $obj = My::Class->new(@_);

Blesses an anonymous hash reference into the given class which inherits 
from C<< Class::Lite >>. Passes all its args to C<< init() >>. 

=head2 init()

    my $obj = $old->init(@_);

This abstract method does nothing at all and returns its object. 
You may wish to override it in your class. 

=head1 GENERATED METHODS 

Accessor methods are generated for each argument on the use-line. 
They all do just what you'd expect. No validation is done. 

    $self   = $self->put_attr($foo);
    $foo    = $self->get_attr;

Put accessors return the object. Get accessors discard any arguments.

=head1 MULTIPLE INHERITANCE

C<< Class::Lite::import() >> is something of a black magic method; it tinkers in caller's package, create a bridge package (in memory), defines methods. It should probably only be called by C<< use() >> or at least from within a C<< BEGIN >> block; no attempt is made to define its behavior if called otherwise. 

Even at compile-time there are questions raised when your class inherits from both C<< Class::Lite >> and some other superclass: 

    package My::Class;
    use Class::Lite qw| foo bar baz |;              # make get/put accessors
    use parent 'Big::Fat::Super';

If the other superclass is pedestrian and just defines methods for you to 
inherit then there's little likelihood of interaction. If the other 
superclass is also trying to define methods with the same names as 
generated accessors then who can say? So don't do that. 

Diamond inheritance is a special case: 

    package My::Big;
    use Class::Lite qw| big1 big2 big3 |;
    
    package My::Tot;
    # I want to inherit from My::Big but I also want Class::Lite's acc's.
    use My::Big;
    use Class::Lite qw| big3 tot1 tot2 |;

This works, regardless of which superclass is use'd first, even if the 
accessor lists overlap. If the My::Big superclass does funny stuff, though, 
all bets are off. Anybody with a use case is welcome to open an issue. 

=head1 SEE ALSO

L<< Object::Tiny|Object::Tiny >>, L<< Mouse|Mouse >>

=head1 INSTALLATION

This module is installed using L<< Module::Build|Module::Build >>. 

=head1 DIAGNOSTICS

=over

=item C<< Invalid accessor name... >>

You passed something horrible on the use-line. Valid arguments to 
C<< import >> need to be quoted strings and valid Perl identifiers. If you 
have in your class some C<< '-$/' >> attribute (which is a valid hash key) 
then you'll have to write your own accessors for it. You won't be able to 
call them, for example, C<< get_-$/() >>. 

This error will attempt to display the offending argument but may not succeed.

=item C<< Failed to generate (package) >>

Something evil happened while doing the heavy lifting: getting into your 
package, getting into the bridge package, setting up the ISA 
relationships, or defining requested accessors. This should not happen and 
isn't your fault (unless you've tried to inherit recursively). Please make 
a bug report. 

=begin for_later

=item C<< some error message >>

Some explanation. 

=item C<< some error message >>

Some explanation. 

=end for_later

=back

=head1 CONFIGURATION AND ENVIRONMENT

None. 

=head1 DEPENDENCIES

There are no non-core dependencies. 

=begin html

<!--

=end html

L<< version|version >> 0.99 E<10> E<8> E<9>
Perl extension for Version Objects

=begin html

-->

<DL>

<DT>    <a href="http://search.cpan.org/perldoc?version" 
            class="podlinkpod">version</a> 0.99 
<DD>    Perl extension for Version Objects

</DL>

=end html

This module should work with any version of perl 5.8.8 and up. 
However, you may need to upgrade some core modules. 

=head1 INCOMPATIBILITIES

None known.

=head1 BUGS AND LIMITATIONS

This is an early release. Reports and suggestions will be warmly welcomed. 

Please report any issues to: 
L<< https://github.com/Xiong/class-lite/issues >>.

=head1 DEVELOPMENT

This project is hosted on GitHub at: 
L<< https://github.com/Xiong/class-lite >>. 

=head1 THANKS

Adam Kennedy (ADAMK) for L<< Object::Tiny|Object::Tiny >>,  on which much of 
this module's code is based. 

=head1 AUTHOR

Xiong Changnian C<< <xiong@cpan.org> >>

=head1 LICENSE

Copyright (C) 2013 
Xiong Changnian C<< <xiong@cpan.org> >>

This library and its contents are released under Artistic License 2.0:

L<< http://www.opensource.org/licenses/artistic-license-2.0.php >>

=cut







util/bump-hack.pl
#!/usr/bin/env perl
#       bump.pl
#       = Copyright 2010 Xiong Changnian <xiong@cpan.org> =
#       = Free Software = Artistic License 2.0 = NO WARRANTY =

use 5.010001;
use strict;
use warnings;
use Carp;

#~ use List::Util qw(first max maxstr min minstr reduce shuffle sum);
#~ use File::Spec;
use File::Spec::Functions qw(
    catdir
    catfile
    catpath
    splitpath
    curdir
    rootdir
    updir
    canonpath
);
use Cwd;
use Getopt::Long;                       # Process command line options
Getopt::Long::Configure ("bundling");   # enable, for instance, -xyz
use Pod::Usage;                         # Build help text from POD
use Text::Glob qw( glob_to_regex );     # Convert shell glob to Perl regex
use List::MoreUtils qw( any );          # The stuff missing in List::Util
#~ use Perl::Version;              # Parse and manipulate Perl version strings
#-#                             # [<-44 cols to end                78 cols ->]

#~ say '>>>', Perl::Version::REGEX, '<<<';
#~ say '>>>', Perl::Version::MATCH, '<<<';
#~ exit(2);

#~ use Devel::Comments '###';
#~ use Devel::Comments '####';
#~ use Devel::Comments '#####';
#~ use Devel::Comments '###', '#####';

#=========# Pseudo-globals
# Constants

# for Getopt::Long; also used to construct %dispatch table
my @getopt_setup     = qw(
    h+
    usage|use
    help
    man
    
    n|dryrun
    v|verbose+
    b|bump=s
);
#    testrequired=i
#    testoptional:s
#    testnegatable!

my $do_prefix    = q{_do_};         # internal; subroutine prefix
my $ignore_file  = q{.bumpskip};    # items to skip in bumping           
my $di           = q{ } x 4;        # indent diff lines by this much

# Variables

my $getopt_okay     ;               # true if GetOptions() didn't fail
my %options         ;               # will be filled by GetOptions()
my @dispatch_setup  ;               # options processed in this order
my $dispatch        ;               # key is option; value is coderef
my $write_files     = 1;            # if false, don't write any file
#~ my $write_files     = 0;            # locked off during development
my $verbosity       = 0;            # verbosity level 1=some, 2=more
my $bump            ;               # new version; specify as string
my $ignore_regex    ;               # giant regex of alternates
my $recursion_depth = 0;            # safety counter

# matches any version
my $version_qr      = qr/(\d+(?:[.]\d+)*)/;

#    >>>(?x-ism: ( (?i: Revision: \s+ ) | v | )
#                              ( \d+ (?: [.] \d+)* )
#                              ( (?: _ \d+ )? ) )<<<
#    >>>(?x-ism: ^ ( \s* ) (?x-ism: ( (?i: Revision: \s+ ) | v | )
#                              ( \d+ (?: [.] \d+)* )
#                              ( (?: _ \d+ )? ) ) ( \s* ) $ )<<<

## Pseudo-globals

#----------------------------------------------------------------------------#
#---------# Main body

# Construct dispatch table
{
    no strict 'refs';
    #### @getopt_setup
    @dispatch_setup      = @getopt_setup;
    for (@dispatch_setup) {
        s/[[:^alpha:]].*$//;   # rm all from the first non-alpha to end
        $dispatch->{$_}     = \&{ $do_prefix . $_        };
    };
    $dispatch->{fail}       = \&{ $do_prefix . 'fail'    };
    $dispatch->{default}    = \&{ $do_prefix . 'default' };
    #### @dispatch_setup
    #### $dispatch
}

# Get command-line arguments
$getopt_okay = GetOptions(
    \%options,
    @getopt_setup,  
);

# Dispatch options
##### %options
if    ( not $getopt_okay ) {            # Getopt::Long failed
    $dispatch->{fail}->(@ARGV);
} 
elsif ( 
            join ( q{}, @ARGV )  eq q{}  
        and scalar keys %options == 0 
      ) {                               # empty command line
    $dispatch->{default}->();
} 
else  {
    for my $key (@dispatch_setup) {
        if ( exists $options{$key} ) {
            #### $key
            $dispatch->{$key}->( $options{$key} );
        };
    };
};

_note( 1, "Bumping files to version '$bump'." );

# Items to do loop
my @todo            ;
if ( join ( q{}, @ARGV )  eq q{} ) {    # nothing given to do
    _fail("No arguments given in command line to bump.");
}
else {
    @todo           = @ARGV;
};
for (@todo) {
    ### argument: $_
    _alter_item($_);
};


exit(0);
#----------------------------------------------------------------------------#

#---------# Dispatch subroutines

sub _do_fail     { $dispatch->{usage}->(2); exit(1); };     # bad @ARGV

sub _do_default  { $dispatch->{usage}->(2); exit(1); };     # no  @ARGV

sub _do_h {                                 # process -h, -hh, -hhh
    my $level       = shift;
    
    given ($level) {
        when (1) { $dispatch->{usage}->(1); exit(0); }
        when (2) { $dispatch->{help}->();   exit(0); }
        when (3) { $dispatch->{man}->();    exit(0); }
        default  { $dispatch->{help}->();   exit(0); }        
    };
    
    return 0;
}; ## _do_h

sub _do_usage {                             # process --usage|-h
    my $level       = shift;
    my @text        = (
        q{Usage: bump [-n][-v]+ -b <version> [<file>|<folder>]+ },
        q{Type 'bump --help' for more info. },
    );
    given ($level) {
        when (1) { _talk(@text); exit(0); }
        when (2) { _yelp(@text); exit(1); }
        default  { _yelp(@text); exit(1); }
    };
    
    return 0;
}; ## _do_usage

sub _do_help {                              # process --help|-hh
    
    my %setup = (
#           -msg        => '#& bump: ',
            -exitval    => 'noexit',
            -verbose    => 1,
            -output     => \*STDOUT,
    );
    pod2usage(%setup);
    
    return 0;
}; ## _do_help

sub _do_man {                               # process --man|-hhh
    
    my %setup = (
#           -msg        => '#& bump: ',
#           -exitval    => 'noexit',
            -verbose    => 2,
            -output     => \*STDOUT,
    );
    pod2usage(%setup);
    
    return 0;
}; ## _do_man

sub _do_n {                                 # process -n|--dryrun
    $write_files = 0;
    _do_v(1);                   # turn on verbose mode for dry run
    return 1;
}; ## _do_n

sub _do_v {                                 # process -v|--verbose
    my $v       = shift;
    $verbosity += $v;
    return 1;
}; ## _do_v

sub _do_b {                                 # process -b|--bump
    $bump = shift;
    #TODO Validate $bump: new version
    return 1;
}; ## _do_b

#---------# Processing subroutines

sub _alter_item {                           # alter file/folder; recursive
    my $item            = shift;
    #### $item
    my @subitems        ;
    
    # Safety limit on recursion
    if ( $recursion_depth++ > 1000 ) {
        _fail( "Recursion depth safety limit exceeded" );
    };
    
    # Skip stuff that should not be bumped at all, ever
    _blacklist() if not $ignore_regex;      # construct if needed
    if ($item =~ $ignore_regex) {
        _note( 2, "$item -- Skipped." );
        return 1;
    };
    
    my $item_stat       = lstat $item;
    if    ( -d _ ) {                        # directory
        opendir my $dh, $item
            or _fail(  "Failed to opendir $item" );
        @subitems       = readdir $dh
            or _fail(  "Failed to readdir $item" );
        closedir $dh
            or _fail( "Failed to closedir $item" );
        
        for (@subitems) {
            next if (/^[.]/);                   # skip self, parent, dotfiles
            _alter_item( catfile( $item, $_ ) );        # recurse
        };
    } 
    elsif ( -f _ ) {                        # plain file
        _alter_file($item);                     # base case
    } 
    else {                                  # other stuff
        _note( 2, "$item -- Skipped." ); 
    };
    
    return 1;
}; ## _alter_item

sub _alter_file {                           # alter a file
    my $item        = shift;
    #### $item
    my $oldfile     ;
    my $infh        ;
    my $outfh       ;
    my $type        ;               # type of processing
    my $lc          ;               # line counter
    
    # Select files for different kinds of change
    given ($item) {
        when (/\.pm/)       { $type = 'pm-pod' }
        when (/\.pod/)      { $type = 'pod' }
        when (/\.agi/)      { $type = 'agi' }
        when (/README$/)    { $type = 'txt' }
        default             { _note( 2, "$item -- Skipped." ); return 1; }
    };
    _note( 1, "$item: " );
    
    # Rename file to backup and open to write
    if ($write_files) {
        $oldfile    = $item . q{~};
        rename $item, $oldfile;
        open $outfh, '>', $item
                or _fail(   "Failed to open $item for writing ");
    }
    else {
        $oldfile    = $item;        # dry run only
    };
    
    # Copy lines
    open $infh, '<', $oldfile
        or _fail(   "Failed to open $oldfile for reading ");
    while ( my $inline = <$infh> ) {
        $lc++;
        my ( $outline, $altered )       = _alter_line( $inline, $type, $lc );
        if ($write_files) {
            print {$outfh} $outline;
        };
        if ($altered) {
            _diff( $lc, $inline, $outline );
        };
    };
    close $infh
        or _fail(   "Failed to close $oldfile ");
    if ( defined $outfh ) {
        close $outfh
            or _fail(   "Failed to close $item ");
    };
    
    return 1;
}; ## _alter_file

sub _alter_line {
    my $line        = shift;
    my $type        = shift;
    my $lc          = shift;
    my $altered     ;
    
    # Alter lines if...
    given ($type) {
        when (/pm/) {
            if ( $line =~ /VERSION =/ ) {
                $line =~ s{(VERSION)(.*?)$version_qr(.*)}
                          {$1$2$bump$4};
                $altered++;
            }
            else {
                continue;
            };
        }  
        when (/pod/) {
            if ( $line =~ /This document describes .*? version/ ) {
                $line =~ s{$version_qr}
                          {$bump};
                $altered++;
            }
            else {
                continue;
            };
        }  
        when (/agi/) {
            if ( $line =~ /^#=version/ ) {
                $line =~ s{$version_qr}
                          {$bump};
                $altered++;
            }
            else {
                continue;
            };
        }  
        when (/txt/) {
            if ( $line =~ /version/ and $lc == 1) {     # first line only
                $line =~ s{$version_qr}
                          {$bump};
                $altered++;
            }
            else {
                continue;
            };
        }  
        
        default {}
    };
    
    return ( $line, $altered );
}; ## _alter_line

sub _blacklist {        # construct blacklist $ignore_regex; once only
    my @ignore      ;
    open my $fh, '<', $ignore_file
        or _fail(   "Failed to open $ignore_file for reading ");
    while (<$fh>) {
        given ($_) {
            when (/^#/)         { }         # skip comment
            when (/^[\s]*$/)    { }         # skip blank
            default             {
                chomp;
                push @ignore, qr/$_/;
            }
        };
    };
    close $fh
        or _fail(   "Failed to close $ignore_file ");
    #### @ignore
    my $ignore_pat      = join q{|}, @ignore;
    $ignore_regex       = qr/$ignore_pat/;
    #### $ignore_regex
    return 1;
}; ## _blacklist

#---------# Screen output subroutines

sub _diff {
    my $lc          = shift;
    my $inline      = shift;
    my $outline     = shift;
    
    return 0 unless $verbosity;
    
    print $di . (               $lc ) . q{: --- } . $inline;
    print $di . ( q{ } x length $lc ) . q{  +++ } . $outline;
    
    return 1;
};


sub _note {
    my $priority        = shift;
    my $intro           = q{#& };                   # introduce each line
    
    if ( $priority <= $verbosity ) {    # smaller number, higher priority
        my $text        = join qq{\n} . $intro, @_;
        say $intro, $text;
    };
}; ## _note

sub _talk { _say_to_screen( \*STDOUT, @_ ) };                   # ~say

sub _yelp { _say_to_screen( \*STDERR, @_ ) };                   # ~warn

sub _fail { _say_to_screen( \*STDERR, @_ ); croak() };     # ~die

sub _say_to_screen {
    my $outfh       = shift;                        # required
    my $intro       = q{#& };                       # introduce each line
    my $prepend     = join q{}, $intro, $0, q{: };  # prepend to all talk               
       
    my $indent      = qq{\n} . $intro 
                    . q{ } x ( (length $prepend) - (length $intro) )
                    ;
    
    my $first_line  = $prepend . shift;
    unshift @_, $first_line;
    my $text        = join $indent, @_; # and indent remaining lines
    
    # Say it
    say {$outfh} $text;
    
    return 1;
}; ## _say_to_screen

__END__

=head1 NAME

bump - Set version numbers in several files

=head1 SYNOPSIS

    $ bump [-n][-v] -b <version> [<file>|<folder>]+ 

=head1 OPTIONS

=over

=item -b|--bump <version>

    $ bump -b '1.2.3'

Required argument. Version-specifying lines will be bumped to this.

=item [-n|--dryrun]

Show what would be done. Implies --verbose.

=item [-v|--verbose]+

If -v, print each line changed. If -vv, very verbose. 

=item [-h|--usage]

Prints a short usage summary.

=item [-hh|--help]

Prints all available options.

=item [-hhh|--man]

Displays all documentation.

=back

=head1 ARGUMENTS

    $ bump -b '1.2.3' lib foo.pm

Remaining command-line arguments specify files and folders to be bumped. 
Folders will be searched recursively. Symlinks will not be followed. 

=head1 AUTHOR

Xiong Changnian  C<< <xiong@cpan.org> >>

=head1 LICENSE

Copyright (C) 2010 Xiong Changnian C<< <xiong@cpan.org> >>

This library and its contents are released under Artistic License 2.0:

L<http://www.opensource.org/licenses/artistic-license-2.0.php>

=cut



top/Build.PL
use 5.008008;
use strict;
use warnings;
use Module::Build;

my $builder = Module::Build->new(
    module_name         => 'Class::Lite',
    license             => 'artistic_2',
    dist_author         => 'Xiong Changnian <xiong@cpan.org>',
    dist_version_from   => 'lib/Class/Lite.pm',
    configure_requires  => { 
        'Module::Build'             => 0.38, 
    },
    requires            => {
        'perl'                      => 5.008008,
        'version'                   => 0.99,
        'overload'                  => 0,
#~         ''  => 0,
    },
    recommends          => {
        'Devel::Comments'           => '1.1.4',
    },
    build_requires      => {
        'Test::More'                => 0.94,
    },
    add_to_cleanup      => [ 'Class-Lite-*' ],
    meta_merge          => {
        resources           => {
            repository          => 'http://github.com/Xiong/class-lite',
        },
        keywords            => [ qw(
            devel       development 
            
            OO  object-oriented class object method
            constructor accessor accessors
            
            lite lightweight tiny
        ) ],
    },
);

$builder->create_build_script();
exit(0);
